File: 2021/day03/solution.py
from typing import Iterable, Tuple, List

def getLines(filename) -> Iterable[str]:
    with open(filename) as f:
        line = f.readline()
        while line:
            yield line[:-1]
            line = f.readline()

# def parseLine(line: str) -> Tuple[str, int]:
#     parts = line.split(" ")
#     return (parts[0], int(parts[1]))

def main():
    part1TestResult = part1("testinput.txt")
    print(f"part1 test: {part1TestResult}")
    part1Result = part1("input.txt")
    print(f"part1: {part1Result}")

    part2TestResult = part2("testinput.txt")
    print(f"part2 test: {part2TestResult}")
    part2Result = part2("input.txt")
    print(f"part2: {part2Result}")

def part1(filename):
    lines = getLines(filename)
    line = next(lines)
    bitSum = list()
    for char in line:
        if char == "0":
            bitSum.append(-1)
        else:
            bitSum.append(1)
    for line in lines:
        for index, char in enumerate(line):
            if char == "0":
                bitSum[index] -= 1
            else:
                bitSum[index] += 1
    gamma = "0b"
    epsilon = "0b"
    for bs in bitSum:
        if bs > 0:
            gamma = gamma + "1"
            epsilon = epsilon + "0"
        else:
            gamma = gamma + "0"
            epsilon = epsilon + "1"
    gammaInt = int(gamma, 2)
    epsilonInt = int(epsilon, 2)
    product = gammaInt * epsilonInt
    return f"gamma={gamma}, gammaInt={gammaInt}, epsilon={epsilon}, epsilonInt={epsilonInt}, product={product}"

def part2(filename):
    lines = list(getLines(filename))

    def getOxygen(lines: List[str], index: int = 0) -> str:
        if len(lines) == 1:
            return lines[0]
        zeroes = list()
        ones = list()
        for line in lines:
            if line[index] == "0":
                zeroes.append(line)
            else:
                ones.append(line)
        if len(ones) >= len(zeroes):
            return getOxygen(ones, index + 1)
        return getOxygen(zeroes, index + 1)

    def getCO2(lines: List[str], index: int = 0) -> str:
        if len(lines) == 1:
            return lines[0]
        zeroes = list()
        ones = list()
        for line in lines:
            if line[index] == "0":
                zeroes.append(line)
            else:
                ones.append(line)
        if len(zeroes) <= len(ones):
            return getCO2(zeroes, index + 1)
        return getCO2(ones, index + 1)

    oxygen = getOxygen(lines)
    oxygenInt = int(oxygen, 2)
    co2 = getCO2(lines)
    co2Int = int(co2, 2)
    product = oxygenInt * co2Int
    return f"oxygen={oxygen}, oxygenInt={oxygenInt}, co2={co2}, co2Int={co2Int}, product={product}"

if __name__ == "__main__":
    main()

File: 2021/day03/golang/solution.go
package main

import (
    "bufio"
    "log"
    "os"
    "strconv"
    // "strings"
)


func check(err error) {
    if err != nil {
        log.Fatal(err)
    }
}

func getLines(path string) ([]string, error) {
    file, err := os.Open(path)
    defer file.Close()

    if err != nil {
        return nil, err
    }

    scanner := bufio.NewScanner(file)
    lines := make([]string, 0, 3000)
    for scanner.Scan() {
        line := scanner.Text()
        lines = append(lines, line)
    }

    return lines, nil
}

func main() {
    log.SetFlags(log.Lmicroseconds)
    part1("input.txt")
    part2("input.txt")
}

func part1(path string) {
    log.SetPrefix("part1: ")
    lines, err := getLines(path)
    check(err)
    lineLen := len(lines[0])
    var gamma string  // most common bits
    var epsilon string  // least common bits
    ones := make([]int, lineLen)
    zeroes := make([]int, lineLen)
    for _, line := range lines {
        for i, char := range line {
            switch char {
                case '1': ones[i]++
                case '0': zeroes[i]++
            }
        }
    }
    log.Println("\tones:\t", ones)
    log.Println("\tzeroes:\t", zeroes)
    for i, _ := range ones {
        if ones[i] > zeroes[i] {
            gamma += "1"
            epsilon += "0"
        } else {
            gamma += "0"
            epsilon += "1"
        }
    }
    log.Println(gamma, epsilon)
    gamma_i, err := strconv.ParseInt(gamma, 2, 64)
    check(err)
    epsilon_i, err := strconv.ParseInt(epsilon, 2, 64)
    check(err)
    log.Println(gamma_i, epsilon_i)
    solution := gamma_i * epsilon_i
    log.Println("solution:", solution)
}

func part2(path string) {
    log.SetPrefix("part2: ")
    lines, err := getLines(path)
    check(err)
    lineLen := len(lines[0])
    o2 := make([]string, len(lines))  // most common bits
    copy(o2, lines)
    for i := 0; i < lineLen && len(o2) > 1; i++ {
        log.Println("o2:", o2)
        ones := 0
        zeroes := 0
        for _, line := range o2 {
            switch line[i] {
                case '1': ones++
                case '0': zeroes++
            }
        }
        var matchChar byte  // why can't we use a rune here?
        if ones >= zeroes {
            matchChar = '1'
        } else {
            matchChar = '0'
        }
        o2New := make([]string, 0, len(o2))
        for _, line := range o2 {
            if line[i] == matchChar {
                o2New = append(o2New, line)
            }
        }
        o2 = o2New
        if len(o2) == 1 { break }
    }
    log.Println("final o2:", o2)

    co2 := make([]string, len(lines))  // least common bits
    copy(co2, lines)
    for i := 0; i < lineLen && len(co2) > 1; i++ {
        log.Println("co2:", co2)
        ones := 0
        zeroes := 0
        for _, line := range co2 {
            switch line[i] {
                case '1': ones++
                case '0': zeroes++
            }
        }
        var matchChar byte  // why can't we use a rune here?
        if ones >= zeroes {
            matchChar = '0'
        } else {
            matchChar = '1'
        }
        co2New := make([]string, 0, len(co2))
        for _, line := range co2 {
            if line[i] == matchChar {
                co2New = append(co2New, line)
            }
        }
        co2 = co2New
    }
    log.Println("final co2:", co2)
    o2_i, err := strconv.ParseInt(o2[0], 2, 64)
    check(err)
    co2_i, err := strconv.ParseInt(co2[0], 2, 64)
    check(err)
    log.Println("solution:", o2_i * co2_i)
}
